From fd4f4295fc4e6e08528869056cb24c4d084e74b6 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 3 Nov 2021 15:45:31 +0100
Subject: [PATCH 0628/1695] drivers: add counter framework

Adds a counter framework which allows:
Provider:
-Allocate/free a counter device.
-Register a counter device.

Client:
-get a counter
-start counter
-stop counter
-get a counter value
-set/cancel an alarm with callback

This change registers a single counter devices. Further changes
will add support for more counters.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git bd6203abafd6adb726b93989cc3fc4a5fe63433b drivers: add counter framework]
Change-Id: I04dbf11ab20571dfeef66da98631e768caf24a7c
Signed-off-by: Ludovic Barre <ludovic.barre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/200757
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Tested-by: Lionel DEBIEVE <lionel.debieve@foss.st.com>
---
 core/drivers/counter/counter.c | 190 +++++++++++++++++++++++++++++++++
 core/drivers/counter/sub.mk    |   1 +
 core/drivers/sub.mk            |   1 +
 core/include/drivers/counter.h | 118 ++++++++++++++++++++
 4 files changed, 310 insertions(+)
 create mode 100644 core/drivers/counter/counter.c
 create mode 100644 core/include/drivers/counter.h

diff --git a/core/drivers/counter/counter.c b/core/drivers/counter/counter.c
new file mode 100644
index 000000000..635aa87ba
--- /dev/null
+++ b/core/drivers/counter/counter.c
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020, STMicroelectronics - All Rights Reserved
+ * Author(s): Ludovic Barre, <ludovic.barre@st.com> for STMicroelectronics.
+ */
+
+#include <assert.h>
+#include <drivers/counter.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <stdlib.h>
+
+static unsigned int list_lock = SPINLOCK_UNLOCK;
+
+static LIST_HEAD(cnt_dev_list_head, counter_device) cnt_dev_list =
+	LIST_HEAD_INITIALIZER(cnt_dev_list_head);
+
+TEE_Result counter_start(struct counter_device *counter)
+{
+	assert(counter && counter->ops);
+
+	if (!counter->ops->start)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	return counter->ops->start(counter);
+}
+
+/*
+ * some counter device could not be stopped
+ * return TEE_SUCCESS if no stop ops
+ */
+TEE_Result counter_stop(struct counter_device *counter)
+{
+	assert(counter && counter->ops);
+
+	if (!counter->ops->stop)
+		return TEE_SUCCESS;
+
+	return counter->ops->stop(counter);
+}
+
+TEE_Result counter_get_value(struct counter_device *counter, unsigned int *ticks)
+{
+	assert(counter && counter->ops);
+
+	if (!counter->ops->get_value)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	return counter->ops->get_value(counter, ticks);
+}
+
+TEE_Result counter_set_alarm(struct counter_device *counter)
+{
+	struct alarm_cfg *alarm = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+	assert(counter && counter->ops);
+
+	alarm = &counter->alarm;
+
+	if (alarm->is_enabled || !alarm->callback || !counter->ops->set_alarm)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (alarm->ticks > counter->max_ticks)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	res = counter->ops->set_alarm(counter);
+	if (!res)
+		counter->alarm.is_enabled = true;
+
+	return res;
+}
+
+TEE_Result counter_cancel_alarm(struct counter_device *counter)
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	assert(counter);
+
+	if (!counter->ops->cancel_alarm)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	res = counter->ops->cancel_alarm(counter);
+	if (!res)
+		counter->alarm.is_enabled = false;
+
+	return res;
+}
+
+#ifdef CFG_DT
+struct counter_device *fdt_counter_get(const void *fdt, int offs)
+{
+	struct counter_device *cnt_dev = NULL;
+	const fdt32_t *cuint;
+	uint32_t exceptions;
+	int phandle = 0;
+
+	assert(fdt);
+
+	cuint = fdt_getprop(fdt, offs, "counter", NULL);
+	if (!cuint)
+		return NULL;
+
+	phandle = fdt32_to_cpu(*cuint);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+
+	LIST_FOREACH(cnt_dev, &cnt_dev_list, dev_list) {
+		if (cnt_dev->phandle == phandle)
+			break;
+	}
+	if (!cnt_dev)
+		EMSG("No counter device registered for phandle: %d", phandle);
+
+	if (cnt_dev && !cnt_dev->is_used)
+		cnt_dev->is_used = true;
+	else
+		cnt_dev = NULL;
+
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	return cnt_dev;
+}
+#endif
+
+struct counter_device *counter_get_by_name(const char *name)
+{
+	struct counter_device *cnt_dev = NULL;
+	uint32_t exceptions;
+
+	assert(name);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+
+	LIST_FOREACH(cnt_dev, &cnt_dev_list, dev_list) {
+		if (!strcmp(cnt_dev->name, name))
+			break;
+	}
+
+	if (cnt_dev && !cnt_dev->is_used)
+		cnt_dev->is_used = true;
+	else
+		cnt_dev = NULL;
+
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	return cnt_dev;
+}
+
+void counter_put(struct counter_device *counter)
+{
+	uint32_t exceptions;
+
+	assert(counter);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	counter->is_used = false;
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+}
+
+struct counter_device *counter_dev_alloc(void)
+{
+	return calloc(1, sizeof(struct counter_device));
+}
+
+void counter_dev_free(struct counter_device *counter)
+{
+	uint32_t exceptions;
+
+	assert(counter);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	LIST_REMOVE(counter, dev_list);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	free(counter);
+}
+
+TEE_Result counter_dev_register(struct counter_device *counter)
+{
+	uint32_t exceptions;
+
+	assert(counter && counter->name && counter->ops);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	LIST_INSERT_HEAD(&cnt_dev_list, counter, dev_list);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	return TEE_SUCCESS;
+}
diff --git a/core/drivers/counter/sub.mk b/core/drivers/counter/sub.mk
index 58ed95c8a..f7699f3a3 100644
--- a/core/drivers/counter/sub.mk
+++ b/core/drivers/counter/sub.mk
@@ -1 +1,2 @@
+srcs-$(CFG_COUNTER_DRIVER) += counter.c
 srcs-$(CFG_STM32_STGEN) += stm32_stgen.c
diff --git a/core/drivers/sub.mk b/core/drivers/sub.mk
index a6ef8e93a..06d2bdcaf 100644
--- a/core/drivers/sub.mk
+++ b/core/drivers/sub.mk
@@ -86,6 +86,7 @@ srcs-$(CFG_VERSAL_PUF) += versal_puf.c
 srcs-$(CFG_VERSAL_HUK) += versal_huk.c
 srcs-$(CFG_CBMEM_CONSOLE) += cbmem_console.c
 srcs-$(CFG_RISCV_PLIC) += plic.c
+subdirs-y += counter
 subdirs-y += crypto
 subdirs-y += firewall
 subdirs-$(CFG_BNXT_FW) += bnxt
diff --git a/core/include/drivers/counter.h b/core/include/drivers/counter.h
new file mode 100644
index 000000000..8f625cb38
--- /dev/null
+++ b/core/include/drivers/counter.h
@@ -0,0 +1,118 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2020, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef COUNTER_H
+#define COUNTER_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <sys/queue.h>
+#include <tee_api_types.h>
+
+/**
+ * @brief Alarm structure.
+ *
+ * @param callback Callback called on alarm (cannot be NULL).
+ * @param ticks Number of ticks that triggers the alarm (absolute value).
+ * @param priv Private data returned in callback.
+ * @param is_enabled True if alarm is enabled, false otherwise.
+ */
+struct alarm_cfg {
+	void (*callback)(unsigned int ticks, void *priv);
+	unsigned int ticks;
+	void *priv;
+	bool is_enabled;
+};
+
+struct counter_device;
+
+struct counter_ops {
+	TEE_Result (*start)(struct counter_device *counter);
+	TEE_Result (*stop)(struct counter_device *counter);
+	TEE_Result (*get_value)(struct counter_device *counter, unsigned int *ticks);
+	TEE_Result (*set_alarm)(struct counter_device *counter);
+	TEE_Result (*cancel_alarm)(struct counter_device *counter);
+};
+
+/**
+ * @brief Start counter device in free running mode.
+ *
+ * @param name name of device.
+ * @param phandle dt phandle.
+ * @param dev_list list on counter device.
+ * @param is_used True if counter is used (exclusive consumer).
+ * @param ops Operation table of the counter.
+ * @param alarm Alarm configuration of the counter.
+ * @param max_ticks Tick max value supported by the counter.
+ * @param priv Optional private data supplied by driver.
+ */
+struct counter_device {
+	const char *name;
+	int phandle;
+	LIST_ENTRY(counter_device) dev_list;
+	bool is_used;
+	const struct counter_ops *ops;
+	struct alarm_cfg alarm;
+	unsigned int max_ticks;
+	void *priv;
+};
+
+/**
+ * @brief Start counter device in free running mode.
+ */
+TEE_Result counter_start(struct counter_device *counter);
+
+/**
+ * @brief Stop counter device.
+ */
+TEE_Result counter_stop(struct counter_device *counter);
+
+/**
+ * @brief Get current counter value.
+ */
+TEE_Result counter_get_value(struct counter_device *counter, unsigned int *ticks);
+
+/**
+ * @brief Set an alarm.
+ */
+TEE_Result counter_set_alarm(struct counter_device *counter);
+
+/**
+ * @brief Cancel an alarm.
+ */
+TEE_Result counter_cancel_alarm(struct counter_device *counter);
+
+#ifdef CFG_DT
+/**
+ * @brief Parse and lookup a counter referenced by a device node.
+ *
+ * @retval counter device if successful, else 0 on error.
+ */
+struct counter_device *fdt_counter_get(const void *fdt, int offs);
+#endif
+
+/**
+ * @brief get a reference on counter device.
+ *
+ * @retval counter device if successful, else 0 on error.
+ */
+struct counter_device *counter_get_by_name(const char *name);
+
+/**
+ * @brief consumer release the counter device.
+ */
+void counter_put(struct counter_device *counter);
+
+/* API for provider */
+static inline void *counter_priv(struct counter_device *counter)
+{
+	return (void *)counter->priv;
+}
+
+struct counter_device *counter_dev_alloc(void);
+void counter_dev_free(struct counter_device *counter);
+TEE_Result counter_dev_register(struct counter_device *counter);
+
+#endif /* COUNTER_H */
-- 
2.47.2

